artiq/device_db_testing/device_db_generation.py

```diff
@@ -0,0 +1,139 @@
+                            }
+            },
+
+            device_name+"_ttl_io_update": {
+                            "type": "local",
+                            "module": "artiq.coredevice.ttl",
+                            "class": devices_config[device_id]["rtio_channels"]["io_update"]["class"],
+                            "arguments": {
+                                    "channel": devices_config[device_id]["rtio_channels"]["io_update"]["channel_num"]
+                            }
+            },
+
+            device_name+"_cpld": {
+                            "type": "local",
+                            "module": "artiq.coredevice.urukul",
+                            "class": "CPLD",
+                            "arguments": {
+                                "spi_device": device_name+"_spi",
+                                "io_update_device": device_name+"_ttl_io_update",
+                                "refclk": 100e6,
+                                "clk_sel": 1
+                            }
+            }
+    }
+
+    for channel_id in devices_config[device_id]["rtio_channels"]:
+        if channel_id[-1].isdigit():
+            for (channel, custom_name) in custom_info.items():
+                if channel_id[-1] in str(channel):
+                    name = custom_name
+                    break
+            else:
+                if generate_all == True:
+                    name = device_name+"_ch"+channel_id[-1]
+                else:
+                    continue
+
+            urukul.update({
+                device_name+"_ttl_"+channel_id: {
+                                "type": "local",
+                                "module": "artiq.coredevice.ttl",
+                                "class": devices_config[device_id]["rtio_channels"][channel_id]["class"],
+                                "arguments": {
+                                        "channel": devices_config[device_id]["rtio_channels"][channel_id]["channel_num"]
+                        }
+                },
+
+                name: {
+                        "type": "local",
+                        "module": "artiq.coredevice.ad9912",
+                        "class": "AD9912",
+                        "arguments": {
+                                "pll_n": 10,
+                                "chip_select": 4+int(channel_id[-1]),
+                                "cpld_device": device_name+"_cpld",
+                                "sw_device": device_name+"_ttl_"+channel_id
+                        }
+                }
+
+            })
+
+    return urukul
+
+def create_device_db(devices_config, core_device, peripherals, generate_all=True):
+    device_db = dict()
+    device_db.update(core_device)
+
+    #Create entries for devices.
+    for device_id in devices_config:
+        for (device, custom_info) in peripherals.items():
+            if device == device_id:
+                info = custom_info
+                break
+        else:

```

Comment:
Does this code work (I haven't tried it myself yet)? Maybe I'm being daft, but what is that `else` in response to? The whitespace looks off...

Reply:
Okay, scratch that. I didn't know python had that feature...